subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4468,22 +4453,15 @@ static int32_t lwm2m_engine_service(const int64_t timestamp)
 
 int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 {
-	struct observe_node *obs, *tmp;
-	sys_snode_t *prev_node = NULL;
+	struct observe_node *obs;
 	int sock_fd = client_ctx->sock_fd;
 	struct lwm2m_message *msg;
 	size_t i;
 
 	/* Remove observes for this context */
-	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&engine_observer_list, obs, tmp,
+	SYS_SLIST_FOR_EACH_CONTAINER(&client_ctx->observer, obs,
```

Comment:
We should keep `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` as we're removing the objects from the list in the loop (SAFE in this case is not about thread-safety but about safe removal of the list nodes from within the loop). Otherwise you break the loop on the first iteration as you clear the `next` pointer.